practices/appeng/adrs/0001-use-typescript.md

```diff
@@ -0,0 +1,107 @@
+
+### Why type safety?
+
+The decision to use TypeScript assumes a consensus that type safety in a programming language is better than no type safety. Below are some reasons for this:
+
+- A strong static typing system can help to reduce the possibility of bugs at runtime, since type errors are captured earlier, during compilation.
+- Type checking significantly reduces the need to write tests or runtime logic that check for unexpected types, since we can be sure that functions are called with the correct argument types. The scope of these errors are largely narrowed down to dealing with unvalidated user input or external data, such as from a 3rd party API.
+- Type checking can also be used to define whether a variable or field is required or possibly undefined. The [billion-dollar mistake](https://en.wikipedia.org/wiki/Tony_Hoare#Apologies_and_retractions) was due to an insufficienlty expressive type system.
+
+## Decision
+
+Our decision is to declare a general preference to use TypeScript over JavaScript when the opportunity comes up.
+
+To clarify, this is _not_ a decision to use TypeScript whenever possible, including as a server language. Rather, when presented with a scenario where JavaScript might be the natural choice of language, TypeScript should be _strongly_ considered instead.
+
+## Why is this applicable to the practice as a whole
+
+As stated above, many of the projects Truss works on use JavaScript to some extent. This decision is applicable to any such project, with the intention of encouraging high code quality and consistent tooling across engineering practices.
+
+## When to not implement this decision
+
+- If the project has no reason to use JavaScript, there may be no reason to use TypeScript. However, this decision does not preclude the option to use TypeScript as a server language if otherwise appropriate.
+- It can be very challenging and time consuming to introduce TypeScript to an existing JavaScript code base, especially if there are many dependencies and/or integrations that may or may not have TypeScript support. In this case it might not be feasible to spend time configuring TypeScript, although the option of introducing TypeScript incrementally (such as only to new code) should also be considered.
+- TypeScript can have a notoriously high learning curve, and may be too costly to adopt by certain projects with tight timing and/or staffing circumstances. However, that overhead in this situation should still be weighed carefully against the payoff of using TypeScript to prevent errors and reduce the need for writing tests later on.
```

Comment:
```suggestion
- TypeScript can have a notoriously high learning curve, and may be too costly to adopt by certain projects with tight timing and/or staffing circumstances. However, that overhead in this situation should still be weighed carefully against the payoff of using TypeScript to prevent errors and reduce the need for writing certain kinds of tests later on.
```